engineapi/proxy.go

```diff
@@ -19,77 +23,296 @@ import (
 )
 
 const (
-	ProxyPollInterval = 5 * time.Second
+	MaxProxyClientConcurrentInUse = 1000
+	DefaultProxyClientPerImage    = 3
+	MaxProxyClientPerImage        = 3 * DefaultProxyClientPerImage
+	ProxyPollInterval             = 5 * time.Second
+	ProxyClientMaxRetry           = 10
+	ProxyGCMaxRetry               = 3
 )
 
-type EngineClientProxy interface {
-	EngineClient
-	BackupTargetBinaryClient
-
-	IsGRPC() bool
-	Start(*longhorn.InstanceManager, logrus.FieldLogger, *datastore.DataStore) error
-	Stop(string) error
-	Ping() error
-}
-
 type EngineClientProxyHandler struct {
-	Clients map[string]EngineClientProxy
+	handle map[string]EngineClientProxys
 
 	logger logrus.FieldLogger
+	ds     *datastore.DataStore
 
-	ds *datastore.DataStore
+	stopCh chan struct{}
 
 	lock *sync.RWMutex
+
+	clock        clock.Clock
+	resyncPeriod time.Duration
+}
+
+type EngineClientProxys struct {
```

Comment:
Do we need to expose this structure or just in-package use? 

This seems like client pool, can we make a meaningful name instead? like `EngineClientProxyPool` or `EngineClientProxyCache`?

Reply:
Updated.